core/lib/node/examples/main_node.rs

```diff
@@ -0,0 +1,86 @@
+//! An incomplete example of how node initialization looks like.
+//! This example defines a `ResourceProvider` that works using the main node env config, and
+//! initializes a single task with a health check server.
+
+use zksync_config::PostgresConfig;
+use zksync_dal::ConnectionPool;
+use zksync_env_config::FromEnv;
+use zksync_node::{
+    healthcheck::IntoHealthCheckTask,
+    node::ZkSyncNode,
+    resource::{pools::MasterPoolResource, Resource, ResourceProvider},
+    task::{
+        healtcheck_server::HealthCheckTask, metadata_calculator::MetadataCalculatorTask,
+        IntoZkSyncTask,
+    },
+};
+
+/// Resource provider for the main node.
```

Comment:
I think the answer to this question is also related to your main review message.

I think that we aren't building a _general-purpose_ framework like Spring. The use cases here will be somewhat limited, main being:
- Use different config formats (env/toml/yaml).
- Run different sets of tasks (main motivation is to run different components on different machines, like we do now; and probably either run or nor run setup-specific tasks (BWIP for main node, reorg detector for EN)).
- Customize tasks -- and that's where resources come in handy. When we are composing the node, we know exactly what behavior we want.

So it seems that it's OK for us to know all the possible resource definitions beforehand. Obviously, we can have some node-specific resources that are provided implicitly, but generally I don't find it too awful for the end crate to explicitly know what resources it provides. Especially that pretty often these resources require some kind of config that you have to provide anyways.

We lose some flexibility this way, but also what we obtain is that we don't have to deal BOTH with topological sorts and having to rely on the call ordering.

> An important requirement in this case would be to be able abstract the resource type from the implementation (i.e., have it defined something like Arc<dyn MyTrait>).

It is possible right now (see e.g. object store resource implementation), but it is resource provider's responsibility to choose which implementation it will give.